### ISSUE TITLE
Some library metadata is not recognized or is 'null'

### ISSUE THREAD

**Issue by koral--** [mikepenz/AboutLibraries#461]
Follow-up of https://github.com/mikepenz/AboutLibraries/issues/460#issue-544532388 for cases when more than license is missing. Environment is the same.

Some data like author, group, description is either missing or shown as literal `null`.
Some of them may be caused by no data provided by authors.

Affected libraries:
* https://github.com/tommybuonomo/dotsindicator
* https://github.com/zxing/zxing
* https://github.com/ota4j-team/opentest4j
* https://github.com/ChuckerTeam/chucker
* https://github.com/square/okhttp
* https://github.com/junit-team/junit5

Actual results:
![image](https://user-images.githubusercontent.com/3340954/71663883-30104a00-2d57-11ea-81fd-bb2858d4de54.png)
![image](https://user-images.githubusercontent.com/3340954/71664031-b331a000-2d57-11ea-8214-576386e88330.png)
![image](https://user-images.githubusercontent.com/3340954/71664062-d9efd680-2d57-11ea-977b-fcfb9b4bf13c.png)
![image](https://user-images.githubusercontent.com/3340954/71664123-0b68a200-2d58-11ea-92bb-46865de47d00.png)

**Comment by mikepenz** [comment]
fun fact: 

![image](https://user-images.githubusercontent.com/1476232/71664321-e1fc4600-2d58-11ea-9563-5bd8de6a5630.png)


That is their "official" description

**Comment by mikepenz** [comment]
Not sure if this is currently blocking you in a real world scenario (I suppose so) 

There is (in a03) actually already a feature which allows you to "enchant" libraries. 

First use the task to get the identifier for the library not being fully supported (due to lacking POM details)

```
 ./gradlew findLibraries
```

and then find the definition file of the library here: 
https://github.com/mikepenz/AboutLibraries/tree/feature/aboutlibs_next_gen/library-definitions/src/main/res/values

Copy it to your project. Get it's identifier from the file. 
![image](https://user-images.githubusercontent.com/1476232/71664468-5931da00-2d59-11ea-86c9-3dac2ee323cb.png)

And then provide the enchanting information:
https://github.com/mikepenz/AboutLibraries/blob/feature/aboutlibs_next_gen/library/src/main/java/com/mikepenz/aboutlibraries/LibsBuilder.kt#L385

**Comment by mikepenz** [comment]
But it definitely sounds like we need to build up a more advanced mapping for enchanting

**Comment by koral--** [comment]
OK, maybe my intent wasn't so clear.
It is not blocking me and I knew that I can add such info manually. However, I don't want to go that way. It is hard to maintain, eg. if I remove that dependency or something gets changed by the library update I would need to also change resources manually.

I would rather fill issues and expect authors to add/fix relevant metadata.

I haven't deeply checked what is present in POMs and what exact algorithm is used by AboutLibraries to retrieve data, so I first tried filling issue here.

Let me know what else (apart from okhttp) is definitely caused by library authors so I'll fill appropriate issues.

**Comment by mikepenz** [comment]
Oh no no I just wanted to ensure in case if you were not aware and it was blocking then this would have been a possible route. 

But it is definitely on my todo to enhance the situation by adding information which we already have present right now as definition file. 

Or establish a proper way of having whitelists/blacklists for libraries which seem to not fully set the POM information :D 

The site you linked before shows btw. very nicely the POM with its information:
https://search.maven.org/artifact/pl.droidsonroids.foqa/deviceinfo/0.1.13/aar


![image](https://user-images.githubusercontent.com/1476232/71665441-402b2800-2d5d-11ea-90c7-8751ec8bbe43.png)


with v5 I want to make use the most of all information available by default by libraries so no manual approach is needed anymore. also resolving which libs are present and having only those resources in the app is a lot more efficient and does not require any more runtime class path checks :)

**Comment by koral--** [comment]
OK, Cool.

For reference here are first issues/PRs reported upstream:
* https://github.com/square/okhttp/issues/5683
* https://github.com/ChuckerTeam/chucker/pull/171

**Comment by mikepenz** [comment]
thanks a ton :)

**Comment by mikepenz** [comment]
I might introduce a fallback for the description in case it is null or empty or so.

Perhaps the URL to the repository, or something like that

**Comment by koral--** [comment]
👍 Sounds good.

Next part of investigation:
* opentest4j has proper name in POM: https://search.maven.org/artifact/org.opentest4j/opentest4j/1.2.0/jar so invisible name seem to be caused by UI issue in About Libraries. There are a lot of authors and perhaps no space for library name and only padding is visible or something like that.

**Comment by mikepenz** [comment]
Partially fixed the long names :) 

Will work more on the feature to add fallback descriptions via the already existing definition files :)

**Comment by koral--** [comment]
In case of zxing-core its POM does not contain license directly. However, it has reference to parent, which contains license and group id.
It looks like gradle plugin should parse `<parent>` tag and take relevant inherited fields from it, maybe even recursively :thinking: 

Chucker has approved my PR.

`dotsindicator` is also mentioned in #460 so I'll comment on it there.

**Comment by mikepenz** [comment]
@koral-- not sure if it is possible to access the parents data. I will have to see. the POM is retrieved from gradle directly for that matter it is not downloaded but directly resolved from the local caches which makes it blazing fast. 
and I am not sure if the parent pom is also retrieved by it. 

Awesome news! :D 

ok :)

**Comment by mikepenz** [comment]
@koral-- so I tried to look into resolving the parent POM. apparently there are kind of 'hacky' approaches to it, but I am not able to make it run. 
In case you are curious a try on it can be found here: https://github.com/mikepenz/AboutLibraries/tree/feature/try_resolve_parent_pom 
including the exception which is thrown. 

Not sure if it is incompatiblity with Gradle 6.1.0 or something else

**Comment by koral--** [comment]
OK, I'll take a look.

**Comment by mikepenz** [comment]
ok I think I have a working concept. 

Please give me 30 minutes to try it out a bit more :)

**Comment by mikepenz** [comment]
@koral-- released 8.0.0-a05 with parent pom resolval logic and some other minor fixes.

**Comment by koral--** [comment]
I've just checked that version and the following issues left:
* gson is missing version and description
`version` tag is nested in `parent`: https://github.com/google/gson/blob/master/gson/pom.xml
description is located in parent POM: https://github.com/google/gson/blob/master/pom.xml
* junit5 artifacts are missing licenses
It seems that they uses EPL:
```xml
    <license>
      <name>Eclipse Public License v2.0</name>
      <url>https://www.eclipse.org/legal/epl-v20.html</url>
    </license>
```
* fabric/crashlytics artifacts are missing licenses
Here is a fragment of the POM:
```xml
    <license>
      <name>Crashlytics Terms of Service</name>
      <url>http://try.crashlytics.com/terms/terms-of-service.pdf</url>
      <distribution>repo</distribution>
    </license>
```
BTW that link redirects via 301 to https://firebase.google.com/products/crashlytics?utm_source=crashlytics_marketing&utm_medium=redirect&utm_campaign=crashlytics_redirect

**Comment by mikepenz** [comment]
I think the next step is also to move the licenses away from the core and only include the licenses which are actually used, as the licenses also have quite some size, and no need to include some which won't be shown

**Comment by mikepenz** [comment]
btw. I released v8.0.0-a06 with the added new licenses as you noted above. also with supporting the parent version and parent description in case the child does not have it

**Comment by koral--** [comment]
Yeah, it is impossible to keep and maintain every single type of license in core.

I haven't checked deeply how everything works now under the hood and what exactly is already done.

IMHO (I might be wrong) when gradle plugin is used it should discover all the licenses and generate XML files with strings (possibly even download content from `<url>` tag, or output link if license type is unknown).
With help of Gradle's @Input(File) and @Output annotations it should be done only once unless files have changed.
The most frequently used licenses like Apache, MIT, ASDL etc. should be hardcoded somewhere to speed up the processes.

Regarding the latest release, I confirm that everything reported is now resolved. I've discovered only one new type of license (not listed before, for `com.crashlytics.sdk.android:answers`):
```xml
    <license>
      <name>Answers Terms of Service</name>
      <url>https://answers.io/terms</url>
      <distribution>repo</distribution>
    </license>
```

**Comment by mikepenz** [comment]
What I have in mind, have those libraries in the definitions dependency, and include them in the plugin, so they will only be included in the project if actually needed by the project. 

If you got some more experience with the input and output annotations though, I am happy about any improvements to it.

**Comment by koral--** [comment]
OK, I'll try to look at that and create a PR or separate.

It seems that this one may be closed after `com.crashlytics.sdk.android:answers` case is resolved.

**Comment by mikepenz** [comment]
Ok `8.0.0-a08` is now released too which now also has all licenses in the definition module. 

As such, only licenses which are used will be included. 
Also potential library enchanting is now offered too

**Comment by arekolek** [comment]
So in 8.0.0-rc02 okhttp3, zxing, gson should be resolved properly? Because for me it doesn't work apparently and I get:

```
UNKNOWN LICENSES / ARTIFACTS WITHOUT LICENSE:
com.squareup.retrofit2:retrofit:
com.google.zxing:core:3.4.0
com.squareup.okhttp3:logging-interceptor:
org.codehaus.mojo:animal-sniffer-annotations:
Answers Terms of Service
com.squareup.okhttp3:okhttp:
com.squareup.okio:okio:
BSD 3-clause
CC0
com.squareup.retrofit2:adapter-rxjava2:
org.joda:joda-money:1.0.1
com.squareup.retrofit2:converter-gson:
net.zetetic:android-database-sqlcipher:4.0.1
org.slf4j:slf4j-api:
com.google.code.gson:gson:
com.google.guava:listenablefuture:9999.0-empty-to-avoid-conflict-with-guava
com.squareup.retrofit2:converter-scalars:
com.google.guava:guava:
com.google.guava:failureaccess:1.0.1
```

Here's the gradle build scan if you wanted to take a look at the dependencies or full console output or whatever https://scans.gradle.com/s/ajunnawughf6q

**Comment by mikepenz** [comment]
thanks for the report @arekolek we will look into those and see to add them to the set of libraries we have to manually improve :/

**Comment by mikepenz** [comment]
@arekolek v8.0.0-rc03 should bring some improvements. 

Answers TOS won't be supported out of the box right now. as (as far as I am aware of, got discontinued in favor of firebase) 

please provide a new export with rc03 to see what may now be missing
